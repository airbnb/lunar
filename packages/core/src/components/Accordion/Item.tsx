import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import ExpandableIcon from '../ExpandableIcon';

export type Props = {
  /** Apply a border. */
  bordered?: boolean;
  /** Content to render if the accordion item is expanded. */
  children?: React.ReactNode;
  /** Whether the accordion item is expanded or not. */
  expanded?: boolean;
  /** Unique id of the accordion item. */
  id: string;
  /** Index amongst a collection of accordion items. */
  index?: number;
  /** Title of the accordion item. */
  title?: React.ReactNode;
  /** Callback fired when the accordion item is clicked. */
  onClick?: (index: number) => void;
};

/**
 * A single accordion item. To be rendered amongst a collection of accordion items, within an accordion.
 */
export class AccordionItem extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    bordered: false,
    children: null,
    expanded: false,
  };

  private handleClick = () => {
    if (this.props.onClick) {
      this.props.onClick(this.props.index!);
    }
  };

  render() {
    const { cx, bordered, children, expanded, id, styles, theme, title } = this.props;

    return (
      <div className={cx(styles.item, bordered && styles.item_bordered)}>
        <button
          className={cx(styles.title)}
          aria-controls={`accordion-body-${id}`}
          aria-selected={expanded}
          id={`accordion-title-${id}`}
          onClick={this.handleClick}
          role="tab"
          tabIndex={0}
          type="button"
        >
          {title && <span className={cx(styles.titleText)}>{title}</span>}

          <ExpandableIcon expanded={!!expanded} size={theme!.unit * 3} />
        </button>

        <section
          className={cx(styles.body, expanded && styles.body_expanded)}
          aria-hidden={!expanded}
          aria-labelledby={`accordion-title-${id}`}
          id={`accordion-body-${id}`}
          role="tabpanel"
        >
          {children}
        </section>
      </div>
    );
  }
}

export default withStyles(
  ({ color, pattern, ui, unit }) => ({
    body: {
      display: 'none',
      padding: `${unit}px ${unit * 2}px ${unit * 2}px`,
    },

    body_expanded: {
      display: 'block',
    },

    item: {
      background: color.accent.bg,
    },

    item_bordered: {
      borderTop: ui.border,
    },

    title: {
      ...pattern.resetButton,
      display: 'flex',
      alignItems: 'center',
      padding: unit * 2,
      textAlign: 'left',
      width: '100%',
    },

    titleText: {
      flex: '1',
    },
  }),
  {
    passThemeProp: true,
  },
)(AccordionItem);
