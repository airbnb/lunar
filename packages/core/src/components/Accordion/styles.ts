import { StyleSheet } from '../../hooks/useStyles';

export const styleSheet: StyleSheet = ({ color }) => ({
  container: {
    borderBottom: '1px solid transparent',
    borderTop: '1px solid transparent',

    ':empty': {
      display: 'none',
    },
  },

  container_bordered: {
    borderBottomColor: color.accent.border,
  },
});

export const styleSheetItem: StyleSheet = ({ pattern, ui, unit }) => ({
  body: {
    display: 'none',
    padding: `${unit}px ${unit * 2}px ${unit * 2}px`,
  },

  body_noSpacing: {
    padding: `0 0 ${unit * 2}px`,
  },

  body_expanded: {
    display: 'block',
  },

  item_bordered: {
    borderTop: ui.border,
  },

  title: {
    ...pattern.resetButton,
    display: 'flex',
    alignItems: 'center',
    padding: unit * 2,
    textAlign: 'left',
    width: '100%',
  },

  title_noSpacing: {
    paddingLeft: 0,
    paddingRight: 0,
  },

  titleText: {
    flex: '1',
  },
});
