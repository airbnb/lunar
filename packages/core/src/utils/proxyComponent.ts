/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';

export default function proxyComponent<T>(
  Component: React.ComponentType<any>,
  implementation: (props: T) => React.ReactElement<any>,
): React.ComponentType<T> {
  function Proxy(props: T) {
    return implementation(props);
  }

  Proxy.displayName = `Proxy(${Component.displayName || Component.name})`;

  return Proxy;
}
