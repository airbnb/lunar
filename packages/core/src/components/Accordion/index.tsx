import React from 'react';
import uuid from 'uuid/v4';
import childrenWithComponentName from '../../prop-types/childrenWithComponentName';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';
import Item, { Props as AccordionItemProps } from './Item';

export type Props = {
  /** Apply a border. */
  bordered?: boolean;
  /** Accordion items and their content. */
  children: NonNullable<React.ReactNode>;
  /** Index of accordion expanded by default. Provide `-1` to collapse all initially. */
  defaultIndex?: number;
};

export type State = {
  id: string;
  index: number;
};

/** A controller for multiple accordion items. */
export class Accordion extends React.Component<Props & WithStylesProps, State> {
  static propTypes = {
    children: childrenWithComponentName('AccordionItem').isRequired,
  };

  static defaultProps = {
    bordered: false,
    defaultIndex: 0,
  };

  state = {
    id: uuid(),
    index: this.props.defaultIndex || 0,
  };

  componentDidUpdate(prevProps: Props) {
    if (this.props.defaultIndex !== prevProps.defaultIndex) {
      this.setState({
        index: this.props.defaultIndex!,
      });
    }
  }

  private handleClick = (index: number) => {
    this.setState(prevState => ({
      index: index === prevState.index ? -1 : index,
    }));
  };

  render() {
    const { bordered, children, styles } = this.props;
    const { id, index } = this.state;

    return (
      <div {...css(styles.container, bordered && styles.container_bordered)} role="tablist">
        {React.Children.map(children, (child, i) => {
          if (!child) {
            return null;
          }

          return React.cloneElement(child as React.ReactElement<AccordionItemProps>, {
            bordered,
            expanded: i === index,
            id: `${id}-${i}`,
            index: i,
            onClick: this.handleClick,
          });
        })}
      </div>
    );
  }
}

export { Item };

export default withStyles(({ color }) => ({
  container: {
    background: color.accent.bg,
    borderBottom: '1px solid transparent',
    borderTop: '1px solid transparent',

    ':empty': {
      display: 'none',
    },
  },

  container_bordered: {
    borderBottomColor: color.accent.border,
  },
}))(Accordion);
