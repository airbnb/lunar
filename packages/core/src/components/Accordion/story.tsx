import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Text from '../Text';
import Accordion, { Item } from '.';

export default {
  title: 'Core/Accordion',
  parameters: {
    inspectComponents: [Accordion, Item],
  },
};

export function multipleItemsWithBorders() {
  return (
    <Accordion bordered>
      <Item title="Item 1" id="one">
        <Text>
          <LoremIpsum />
        </Text>
      </Item>

      <Item title="Item 2" id="two">
        <Text>
          <LoremIpsum />
        </Text>
      </Item>

      <Item title="Item 3" id="three">
        <Text>
          <LoremIpsum />
        </Text>
      </Item>
    </Accordion>
  );
}

multipleItemsWithBorders.story = {
  name: 'Multiple items with borders.',
};

export function singleItemInitiallyClosed() {
  return (
    <Accordion defaultIndex={-1}>
      <Item title="Item 1" id="one">
        <Text>
          <LoremIpsum />
        </Text>
      </Item>
    </Accordion>
  );
}

singleItemInitiallyClosed.story = {
  name: 'Single item initially closed.',
};

export function customTitleComponent() {
  return (
    <Accordion bordered>
      <Item
        title={
          <header>
            <div>Main Title</div>
            <small>Subtitle</small>
          </header>
        }
        id="one"
      >
        <Text>
          <LoremIpsum />
        </Text>
      </Item>
    </Accordion>
  );
}

customTitleComponent.story = {
  name: 'Custom title component.',
};

export function withNoSpacing() {
  return (
    <Accordion bordered>
      <Item noSpacing title="Item 1" id="one">
        <Text>
          <LoremIpsum />
        </Text>
      </Item>

      <Item noSpacing title="Item 2" id="two">
        <Text>
          <LoremIpsum />
        </Text>
      </Item>
    </Accordion>
  );
}

withNoSpacing.story = {
  name: 'With no spacing.',
};
