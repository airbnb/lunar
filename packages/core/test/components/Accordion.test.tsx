import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import { unwrapHOCs } from '@airbnb/lunar-test-utils';
import Accordion from '../../src/components/Accordion';
import AccordionItem, { Props as AccordionItemProps } from '../../src/components/Accordion/Item';
import proxyComponent from '../../src/utils/proxyComponent';

function unwrap(element: any): Enzyme.ShallowWrapper {
  return unwrapHOCs(shallow(element), 'Accordion', {}, { render: true });
}

describe('<Accordion />', () => {
  it('errors if non-accordion item children are passed', () => {
    expect(() => unwrap(<Accordion>Foo</Accordion>)).toThrowError();
  });

  it('does not error if a proxyComponent wrapping a AccordionItem is passed', () => {
    const ProxiedAccordionItem = proxyComponent(AccordionItem, (props: AccordionItemProps) => (
      <AccordionItem {...props} />
    ));

    expect(() =>
      unwrap(
        <Accordion>
          <ProxiedAccordionItem id="0" title="Label" />
        </Accordion>,
      ),
    ).not.toThrowError();
  });

  it('errors if proxyComponent wrapping a non-AccordionItem is passed', () => {
    const IncompatibleComponent = () => <div />;
    const ProxiedNonAccordionItem = proxyComponent(IncompatibleComponent, () => <div />);

    expect(() =>
      unwrap(
        <Accordion>
          <ProxiedNonAccordionItem />
        </Accordion>,
      ),
    ).toThrowError();
  });

  it('renders expected number of accordion items', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={1}>
        <AccordionItem id="0" title="One" />
        <AccordionItem id="1" title="Two" />
        <AccordionItem id="2" title="Three" />
      </Accordion>,
    );

    expect(wrapper.find(AccordionItem)).toHaveLength(3);
  });

  it('renders bordered', () => {
    const wrapper = shallow(
      <Accordion bordered>
        <AccordionItem id="0" title="Label" />
      </Accordion>,
    ).dive();

    // id is generated with uuid, need to manually set it for snapshots to match
    wrapper.setState({
      id: 0,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('has accessibility role set', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={3}>
        <AccordionItem id="0" title="Label" />
      </Accordion>,
    );

    expect(wrapper.prop('role')).toBe('tablist');
  });

  it('sets index state using `defaultIndex`', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={3}>
        <AccordionItem id="0" title="Label" />
      </Accordion>,
    );

    expect(wrapper.state('index')).toBe(3);
  });

  it('sets expanded state to accordion item by index', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={1}>
        <AccordionItem id="0" title="One" />
        <AccordionItem id="1" title="Two" />
        <AccordionItem id="2" title="Three" />
      </Accordion>,
    );

    expect(
      wrapper
        .find(AccordionItem)
        .at(0)
        .prop('expanded'),
    ).toBe(false);

    expect(
      wrapper
        .find(AccordionItem)
        .at(1)
        .prop('expanded'),
    ).toBe(true);

    expect(
      wrapper
        .find(AccordionItem)
        .at(2)
        .prop('expanded'),
    ).toBe(false);
  });

  it('adds indices to items', () => {
    const wrapper = unwrap(
      <Accordion>
        <AccordionItem id="0" title="One" />
        <AccordionItem id="1" title="Two" />
        <AccordionItem id="2" title="Three" />
      </Accordion>,
    );

    expect(
      wrapper
        .find(AccordionItem)
        .at(0)
        .prop('index'),
    ).toBe(0);

    expect(
      wrapper
        .find(AccordionItem)
        .at(1)
        .prop('index'),
    ).toBe(1);

    expect(
      wrapper
        .find(AccordionItem)
        .at(2)
        .prop('index'),
    ).toBe(2);
  });

  it('sets expanded state to false for all items if defaultIndex is negative', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={-1}>
        <AccordionItem id="0" title="One" />
        <AccordionItem id="1" title="Two" />
        <AccordionItem id="2" title="Three" />
      </Accordion>,
    );

    expect(
      wrapper
        .find(AccordionItem)
        .at(0)
        .prop('expanded'),
    ).toBe(false);

    expect(
      wrapper
        .find(AccordionItem)
        .at(1)
        .prop('expanded'),
    ).toBe(false);

    expect(
      wrapper
        .find(AccordionItem)
        .at(2)
        .prop('expanded'),
    ).toBe(false);
  });

  it('handles falsey items', () => {
    const wrapper = unwrap(
      <Accordion>
        {false && <AccordionItem id="0" title="One" />}
        {null && <AccordionItem id="0" title="Two" />}
        <AccordionItem id="0" title="Three" />
      </Accordion>,
    );

    expect(wrapper.find(AccordionItem)).toHaveLength(1);
  });

  it('updates state index when `defaultIndex` changes', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={3}>
        <AccordionItem id="0" title="Label" />
      </Accordion>,
    );

    expect(wrapper.state('index')).toBe(3);

    wrapper.setProps({
      defaultIndex: 1,
    });

    expect(wrapper.state('index')).toBe(1);
  });

  it('updates index when `handleClick` is triggered', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={3}>
        <AccordionItem id="0" title="Label" />
      </Accordion>,
    );

    expect(wrapper.state('index')).toBe(3);

    wrapper
      .find(AccordionItem)
      .at(0)
      .prop('onClick')!(0);

    expect(wrapper.state('index')).toBe(0);
  });

  it('updates index to -1 when current index is clicked', () => {
    const wrapper = unwrap(
      <Accordion defaultIndex={0}>
        <AccordionItem id="0" title="Label" />
      </Accordion>,
    );

    expect(wrapper.state('index')).toBe(0);

    wrapper
      .find(AccordionItem)
      .at(0)
      .prop('onClick')!(0);

    expect(wrapper.state('index')).toBe(-1);
  });
});
