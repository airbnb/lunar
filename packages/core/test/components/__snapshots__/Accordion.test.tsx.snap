// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> renders bordered 1`] = `
<div
  className="container_1ytewkj-o_O-container_bordered_w9a1au"
  role="tablist"
>
  <AccordionItem
    bordered={true}
    expanded={true}
    id="0-0"
    index={0}
    key=".0"
    onClick={[Function]}
    title="Label"
  />
</div>
`;
